wgpu-core/src/command/mod.rs

```diff
@@ -78,21 +78,185 @@ pub(crate) enum CommandEncoderStatus {
     ///
     /// [`Global::queue_submit`] drops command buffers unless they are
     /// in this state.
-    Finished,
+    Finished(CommandBufferMutable),
 
     /// An error occurred while recording a compute or render pass.
     ///
     /// When a `CommandEncoder` is left in this state, we have also
     /// returned an error result from the function that encountered
     /// the problem. Future attempts to use the encoder (for example,
-    /// calls to [`CommandBufferMutable::check_recording`]) will also return
-    /// errors.
-    ///
-    /// Calling [`Global::command_encoder_finish`] in this state
-    /// discards the command buffer under construction.
+    /// calls to [`Self::record`]) will also return errors.
     Error,
 }
 
+/// Changes the state of the given encoder.
+///
+/// Takes `&mut CommandEncoderStatus`, the previous state kind and the next state kind.
+///
+/// Returns `&mut CommandBufferMutable`.
+///
+/// Will panic if the previous state is not accurate.
+macro_rules! change_state_with_inner {
```

Comment:
Now that I've read a little more, I'm not sure this macro is doing anything valuable:

- It is used only at points where we've already matched the state.

- We never actually use the `&mut CommandBufferMutable` that it produces.

At the very least, we should remove the code that rematches to get the `&mut CommandBufferMutable`, since we're just using the `Deref` and `DerefMut` impls on `EncoderGuard` instead.

Reply:
> At the very least, we should remove the code that rematches to get the `&mut CommandBufferMutable`, since we're just using the `Deref` and `DerefMut` impls on `EncoderGuard` instead.

This we can do, I think it was returning `&mut CommandBufferMutable` because that was required in a previous version of the patch (pre `EncoderGuard`).

Reply:
I removed the macro.